src/Parsers/ASTColumnsTransformers.cpp

```diff
@@ -90,27 +91,52 @@ void ASTColumnsExceptTransformer::formatImpl(const FormatSettings & settings, Fo
         settings.ostr << ")";
 }
 
-void ASTColumnsExceptTransformer::transform(ASTs & nodes) const
+void ASTColumnsExceptTransformer::transform(ASTs & nodes, std::vector<String> databases) const
 {
     std::set<String> expected_columns;
     for (const auto & child : children)
-        expected_columns.insert(child->as<const ASTIdentifier &>().name());
+    {
+        const auto & ident = child->as<const ASTIdentifier &>();
+        expected_columns.insert(ident.name());
+    }
+
+    bool have_database_name = databases.size() > 0 ? true : false;
+    std::vector<String>::iterator db = databases.begin();
 
     for (auto it = nodes.begin(); it != nodes.end();)
     {
         if (const auto * id = it->get()->as<ASTIdentifier>())
         {
-            auto expected_column = expected_columns.find(id->shortName());
+            String full_name = id->name();
+            // match table.column pattern
+            auto expected_column = expected_columns.find(full_name);
+
+            // match column pattern
+            if (expected_column == expected_columns.end())
+                expected_column = expected_columns.find(id->shortName());
+
+            // match db.table.column pattern
+            if (expected_column == expected_columns.end() && have_database_name)
+                expected_column = expected_columns.find(*db + "." + full_name);
+
             if (expected_column != expected_columns.end())
             {
                 expected_columns.erase(expected_column);
                 it = nodes.erase(it);
             }
             else
+            {
                 ++it;
+                if (have_database_name)
+                    ++db;
```

Comment:
I still have some difficulties understanding this code.
Why moving db iterator is correct? What will happen if it will reach databases.end()?
Why concatenating db and table name with `.` is correct?